added document for testing with nodejs (#227)
solana-foundation/developer-content: docs/programs/testing.md

---
title: "Testing with NodeJS"
description: "Testing native solana programs written with rust using NodeJS"
---

When developing programs on Solana, ensuring their correctness and reliability

> is crucial. Until now devs have been using solana-test-validator for testing.

Reviewer: ```suggestion
is crucial. Until now devs have been using `solana-test-validator` for testing.
```

> This document covers testing your Solana program with NodeJS
> using `solana-bankrun`.

Reviewer: ```suggestion
This document covers testing your Solana program with Node.js
using `solana-bankrun`.
```

## Overview

There are two ways to test programs on Solana:

1. [solana-test-validator](https://docs.solanalabs.com/cli/examples/test-validator):
   That spins up a local emulator of the Solana Blockchain on your local machine
   which receives the transactions to be processed by the validator.

> 2. The various BanksClient-based test frameworks for SBF(Solana Bytecode Format)

Reviewer: Can you add a link for BanksClient, or tell readers what BanksClient is?

   programs: Bankrun is a framework that simulates a Solana bank’s operations,
   enabling developers to deploy, interact with, and assess the behaviour of
   programs under test conditions that mimic the mainnet. It helps set up the
   test environment and offers tools for detailed transaction insights,
   enhancing debugging and verification. With the client, we can load programs,
   and simulate and process transactions seamlessly.
   [solana-program-test](https://docs.rs/solana-program-test) (Rust),
   [solana-bankrun](https://github.com/kevinheavey/solana-bankrun) (Rust,
   JavaScript), [anchor-bankrun](https://www.npmjs.com/package/anchor-bankrun)
   (Anchor, JavaScript),
   [solders.bankrun](https://kevinheavey.github.io/solders/api_reference/bankrun.html)
   (Python) are examples of the BanksClient-based testing framework.

>    ```
>    Note:
> 
>    [pnpm create solana-program](https://github.com/solana-program/create-solana-program) can help you generate JS and Rust clients including tests.
>    Anchor is not yet supported.
>    ```

Reviewer: ```suggestion
> [`pnpm create solana-program`](https://github.com/solana-program/create-solana-program) can help you generate JS and Rust clients including tests.
> Anchor is not yet supported.
```

Reviewer: See https://github.com/solana-foundation/developer-content/blob/main/CONTRIBUTING.md#callout

In this guide, we are using Solana Bankrun. `Bankrun` is a superfast, powerful,
and lightweight framework for testing Solana programs in Node.js.

> - The biggest advantage of using Solana Bankrun is that you don’t have to set up
>   an environment to test programs like you’d have to do while using the
>   solana-test-validator. Instead, you can do that with a piece of code, inside
>   the tests.
> - It also dynamically sets time and account data, which isn’t possible with
>   solana-test-validator

Reviewer: ```suggestion
- The biggest advantage of using Solana Bankrun is that you don’t have to set up
  an environment to test programs like you’d have to do while using the
  `solana-test-validator`. Instead, you can do that with a piece of code, inside
  the tests.
- It also dynamically sets time and account data, which isn’t possible with
  `solana-test-validator`
```

## Installation

Add `solana-bankrun` as a dev dependency to your node project. If your Solana
program is not a node project yet, you can initialize it using `npm init`.

```bash
npm i -D solana-bankrun
```

## Usage

### Program Directory

Firstly, the program's `.so` file must be present in one of the following
directories:

- `./tests/fixtures` (just create this directory if it doesn't exist already).
- Your current working directory.
- A directory you define in the `BPF_OUT_DIR` or `SBF_OUT_DIR` environment
  variables. `export BPF_OUT_DIR=’/path/to/binary’`
- Build your program specifying the correct directory so that library can pick
  the file up from directory just from the name.
  `cargo build-sbf --manifest-path=./program/Cargo.toml --sbf-out-dir=./tests/fixtures`

### Testing Framework

solana-bankrun is used in JavaScript or TypeScript with testing frameworks

> like [ts-mocha](https://www.npmjs.com/package/ts-mocha), [ava,](https://github.com/avajs/ava) [Jest](https://jestjs.io/),
> etc. Make sure to get started with any of the above.

Reviewer: ```suggestion
like [ts-mocha](https://www.npmjs.com/package/ts-mocha), [ava](https://github.com/avajs/ava), [Jest](https://jestjs.io/),
etc. Make sure to get started with any of the above.
```

> Add a [npm script](https://docs.npmjs.com/cli/v9/using-npm/scripts) to test your

Reviewer: ```suggestion
Add an [npm script](https://docs.npmjs.com/cli/v9/using-npm/scripts) to test your
```

deploy programs and add accounts as instructed.

```typescript
import { start } from "solana-bankrun";
import { PublicKey } from "@solana/web3.js";

test("testing program instruction", async () => {
  const programId = PublicKey.unique();
  const context = await start([{ name: "program_name", programId }], []);

  const client = context.banksClient;
  const payer = context.payer;
  // write tests
});
```

### Bankrun `context`

- We get access to the Bankrun `context` from the `start` function. Context
  contains a BanksClient, a recent blockhash and a funded payer keypair.
- `context` has a `payer`, which is a funded keypair that can be used to sign
  transactions.
- `context` also has `context.lastBlockhash` or `context.getLatestBlockhash` to
  make fetching [Blockhash](https://solana.com/docs/terminology#blockhash)
  convenient during tests.
- `context.banksClient` is used to send transactions and query account data from
  the ledger state. For example, sometimes
  [Rent](https://solana.com/docs/terminology#rent) (in lamports) is  
  required to build a transaction to be submitted, for example, when using the

>   SystemProgram's  
>   creates account instruction. You can do that using BanksClient:

Reviewer: ```suggestion
  SystemProgram's  
  createAccount() instruction. You can do that using BanksClient:
```

  ```typescript
  const rent = await client.getRent();

  const Ix: TransactionInstruction = SystemProgram.createAccount({
    // ...
    lamports: Number(rent.minimumBalance(BigInt(ACCOUNT_SIZE))),
    //....
  });
  ```

- You can read account data from BanksClient using `getAccount` function
  ```typescript
  AccountInfo = await client.getAccount(counter);
  ```

### Process Transaction

> `processTransaction` function executes the transaction with the loaded programs
> and accounts from the start function and will return a transaction object.

Reviewer: ```suggestion
The `processTransaction()` function executes the transaction with the loaded programs
and accounts from the start function and will return a transaction.
```

   That spins up a local emulator of the Solana Blockchain on your local machine
   which receives the transactions to be processed by the validator.
2. The various
   [BanksClient-based](https://docs.rs/solana-banks-client/latest/solana_banks_client/)
   test frameworks for SBF (Solana Bytecode Format) programs: Bankrun is a
   framework that simulates a Solana bank’s operations, enabling developers to
   deploy, interact with, and assess the behaviour of programs under test
   conditions that mimic the mainnet. It helps set up the test environment and
   offers tools for detailed transaction insights, enhancing debugging and
   verification. With the client, we can load programs, and simulate and process
   transactions seamlessly.
   [solana-program-test](https://docs.rs/solana-program-test) (Rust),
   [solana-bankrun](https://github.com/kevinheavey/solana-bankrun) (Rust,
   JavaScript), [anchor-bankrun](https://www.npmjs.com/package/anchor-bankrun)
   (Anchor, JavaScript),
   [solders.bankrun](https://kevinheavey.github.io/solders/api_reference/bankrun.html)
   (Python) are examples of the BanksClient-based testing framework.

   ```

>    Note:

Reviewer 2: @mikemaccana @heyAyushh is this code block intentional?

Reviewer: Nope. Ayush tried to do a code block, I converted it to a quote and linked the docs but evidently left some backticks in (suggestions have their own backticks so it was probably that).  I'll fix it now.

https://github.com/solana-foundation/developer-content/pull/227#discussion_r1701978151

Reviewer: https://github.com/solana-foundation/developer-content/pull/292

Author: Thanks @mikemaccana, sorry it slipped

---
title: "Testing with NodeJS"
description: "Testing native solana programs written with rust using NodeJS"
---

When developing programs on Solana, ensuring their correctness and reliability
is crucial. Until now devs have been using solana-test-validator for testing.
This document covers testing your Solana program with NodeJS
using `solana-bankrun`.

## Overview

There are two ways to test programs on Solana:

1. [solana-test-validator](https://docs.solanalabs.com/cli/examples/test-validator):
   That spins up a local emulator of the Solana Blockchain on your local machine
   which receives the transactions to be processed by the validator.

> 2. BanksClient-based test framework for SBF(Solana Bytecode Format) programs:

Reviewer: ```suggestion
2. The various BanksClient-based test frameworks for SBF(Solana Bytecode Format) programs:
```

From the following sentence, it seems like there are more than one BanksClient-based test frameworks.

   Bankrun is a framework that simulates a Solana bank’s operations, enabling
   developers to deploy, interact with, and assess the behaviour of programs
   under test conditions that mimic the mainnet. It helps set up the test
   environment and offers tools for detailed transaction insights, enhancing
   debugging and verification. With the client, we can load programs, and
   simulate and process transactions seamlessly.
   [solana-program-test](https://docs.rs/solana-program-test) (Rust),
   [solana-bankrun](https://github.com/kevinheavey/solana-bankrun) (Rust,
   JavaScript), [anchor-bankrun](https://www.npmjs.com/package/anchor-bankrun)
   (Anchor, JavaScript),
   [solders.bankrun](https://kevinheavey.github.io/solders/api_reference/bankrun.html)

>    (Rust, Python) are examples of the BanksClient-based testing framework.

Reviewer: ```suggestion
   (Rust, Python) are examples of BanksClient-based testing frameworks.
```

   ```
   Note:

   [pnpm create solana-program](https://github.com/solana-program/create-solana-program) can help you generate JS and Rust clients including tests.
   Anchor is not yet supported.
   ```

In this guide, we are using Solana Bankrun. `Bankrun` is a superfast, powerful,

> and lightweight framework for testing Solana programs in NodeJS.

Reviewer: ```suggestion
and lightweight framework for testing Solana programs in Node.js.
```

See https://nodejs.org/ for the proper capitalization.

- The biggest advantage of using Solana Bankrun is that you don’t have to set up
  an environment to test programs like you’d have to do while using the

>   solana-test-validator. Instead, you can do that with a piece of code inside

Reviewer: ```suggestion
  solana-test-validator. Instead, you set up the environment with code, inside
```

  the tests.
- It also dynamically setting time and account data, which isn’t possible with
  solana-test-validator

## Installation

Add `solana-bankrun` as a dev dependency to your node project. If your Solana
program is not a node project yet, you can initialize it using `npm init`.

```bash
npm i -D solana-bankrun
```

## Usage

### Program Directory

Firstly, the program's `.so` file must be present in one of the following

> directories:

Reviewer: Should you add a section on how to download `.so`s from mainnet?

Author: will be good if we add that in future, since the guide suffices for hello world.

  the file up from directory just from the name.
  `cargo build-sbf --manifest-path=./program/Cargo.toml --sbf-out-dir=./tests/fixtures`

### Testing Framework

solana-bankrun is used in JavaScript or TypeScript with testing frameworks
like [ts-mocha](https://www.npmjs.com/package/ts-mocha), [ava,](https://github.com/avajs/ava) [Jest](https://jestjs.io/),
etc. Make sure to get started with any of the above.

Add a [npm script](https://docs.npmjs.com/cli/v9/using-npm/scripts) to test your
program and create your `test.ts` file inside `tests` folder.

```json
{
  "scripts": {
    "test": "pnpm ts-mocha -p ./tsconfig.json -t 1000000 ./tests/test.ts"
  }
}
```

### Start

`start` function from `solana-bankrun` spins up a BanksServer and a BanksClient,
deploy programs and add accounts as instructed.

```tsx
import { start } from "solana-bankrun";
import { PublicKey } from "@solana/web3.js";

test("testing program instruction", async () => {
  const programId = PublicKey.unique();
  const context = await start([{ name: "program_name", programId }], []);

  const client = context.banksClient;
  const payer = context.payer;
  // write tests
});
```

> ### Bankrun Context

Reviewer: ```suggestion
### Bankrun `context`
```

> - Get access to bankrun context from start function. Context contains a
>   BanksClient, a recent blockhash and a funded payer keypair.

Reviewer: ```suggestion
- We get access to the Bankrun `context` from the `start` function.
```

> - The `context` variable has various components that can be used while creating

Reviewer: ```suggestion
```

Vague. Suggest deleting it, the previous sentence says enough.

  a transaction to test your program.

> - `payer` from context object is a funded keypair that can be used to sign

Reviewer: ```suggestion
- `context` has a `payer`, which is a funded keypair that can be used to sign
```

  transactions.

> - [Blockhash](https://solana.com/docs/terminology#blockhash) can be retrieved
>   from context using `context.lastBlockhash` or `context.getLatestBlockhash` to
>   be added in transactions to test.

Reviewer: ```suggestion
- `context` also has `context.lastBlockhash` or `context.getLatestBlockhash` to make fetching [Blockhash](https://solana.com/docs/terminology#blockhash) convenient during tests.
```

> ### BanksClient
> 
> - It’s a client for the ledger state, from the perspective of an arbitrary
>   validator. The client is used to send transactions and query account data,
>   among other things.
> - BanksClient can be retrieved from context using `context.banksClient` which
>   drives the rest of the importance here.
> - Sometimes [Rent](https://solana.com/docs/terminology#rent) (in lamports) is
>   required to build a transaction to be submitted, for example, SystemProgram
>   creates account instruction. You can also do that using BanksClient. Code
>   below

Reviewer: ```suggestion

- `context.banksClient` is used to send transactions and query account data from the ledger state. For example, sometimes [Rent](https://solana.com/docs/terminology#rent) (in lamports) is
  required to build a transaction to be submitted, for example, when using the SystemProgram's
  creates account instruction. You can do that using BanksClient:
```

Reviewer: This doesn't need another heading because it's another feature of `context`
